sway/input/text_input.c

```diff
@@ -26,6 +31,329 @@ static struct sway_text_input *relay_get_focused_text_input(
+		struct wlr_surface *surface) {
+	wl_list_remove(&popup->focused_surface_unmap.link);
+
+	if (surface == NULL) {
+		wl_list_init(&popup->focused_surface_unmap.link);
+		goto update_popup;
+	}
+
+	if (wlr_surface_is_layer_surface(surface)) {
+		struct wlr_layer_surface_v1 *layer_surface =
+			wlr_layer_surface_v1_from_wlr_surface(surface);
+		struct sway_layer_surface *layer =
+			layer_from_wlr_layer_surface_v1(layer_surface);
+		wl_signal_add(
+			&layer->layer_surface->events.unmap, &popup->focused_surface_unmap);
+		goto update_popup;
+	}
+
+	struct sway_view *view = view_from_wlr_surface(surface);
+	wl_signal_add(&view->events.unmap, &popup->focused_surface_unmap);
+
+	// Since the focus has changed, the popup may have to adjust
+update_popup:
+	input_popup_update(popup);
+}
+
+static void handle_im_new_popup_surface(struct wl_listener *listener,
+		void *data) {
+	struct sway_input_method_relay *relay = wl_container_of(listener, relay,
+		input_method_new_popup_surface);
+	struct sway_input_popup *popup = calloc(1, sizeof(*popup));
+	popup->relay = relay;
+	popup->popup_surface = data;
+	popup->popup_surface->data = popup;
+
+	wl_signal_add(&popup->popup_surface->events.map, &popup->popup_map);
+	popup->popup_map.notify = handle_im_popup_map;
+	wl_signal_add(
+		&popup->popup_surface->events.unmap, &popup->popup_unmap);
+	popup->popup_unmap.notify = handle_im_popup_unmap;
+	wl_signal_add(
+		&popup->popup_surface->events.destroy, &popup->popup_destroy);
+	popup->popup_destroy.notify = handle_im_popup_destroy;
+	wl_signal_add(&popup->popup_surface->surface->events.commit,
+		&popup->popup_surface_commit);
+	popup->popup_surface_commit.notify = handle_im_popup_surface_commit;
+	wl_list_init(&popup->focused_surface_unmap.link);
+	popup->focused_surface_unmap.notify = handle_im_focused_surface_unmap;
+
+	struct sway_text_input *text_input = relay_get_focused_text_input(relay);
+	if (text_input != NULL) {
+		input_popup_set_focus(popup, text_input->input->focused_surface);
+	} else {
+		input_popup_set_focus(popup, NULL);
+	}
+
+	wl_list_insert(&relay->input_popups, &popup->link);
+}
+
+bool sway_input_popup_get_position(
+        struct sway_input_popup *popup, int *lx, int *ly) {
+	struct sway_text_input *text_input =
+		relay_get_focused_text_input(popup->relay);
+	if (text_input == NULL || text_input->input->focused_surface == NULL) {
+		*lx = 0;
+		*ly = 0;
+		return false;
+	}
+
+	struct wlr_surface *focused_surface = text_input->input->focused_surface;
+	if (wlr_surface_is_layer_surface(focused_surface)) {
+		struct wlr_layer_surface_v1 *layer_surface =
+			wlr_layer_surface_v1_from_wlr_surface(focused_surface);
+		struct sway_layer_surface *layer =
+			layer_from_wlr_layer_surface_v1(layer_surface);
+		*lx = layer->geo.x + popup->x;
+		*ly = layer->geo.y + popup->y;
+		return true;
+	}
+
+
+	struct sway_view *view = view_from_wlr_surface(
+		text_input->input->focused_surface);
+	if (view->container == NULL) {
```

Comment:
This can return NULL here.

This logic doesn't work on subsurfaces nor xdg_popups.